system/src/system_sleep.cpp

```diff
@@ -251,3 +252,147 @@ int system_sleep_pins(const uint16_t* pins, size_t pins_count, const InterruptMo
     network_connect_cancel(0, 1, 0, 0);
     return system_sleep_pin_impl(pins, pins_count, modes, modes_count, seconds, param, reserved);
 }
+
+// Since sleep 2.0
+#if HAL_PLATFORM_CELLULAR || HAL_PLATFORM_WIFI || HAL_PLATFORM_MESH || HAL_PLATFORM_ETHERNET
+static bool system_sleep_network_suspend(network_interface_index index) {
+    bool resume = false;
+    // Disconnect from network
+    if (network_connecting(index, 0, NULL) || network_ready(index, 0, NULL)) {
+        if (network_connecting(index, 0, NULL)) {
+            network_connect_cancel(index, 1, 0, 0);
+        }
+        network_disconnect(index, NETWORK_DISCONNECT_REASON_SLEEP, NULL);
+        resume = true;
+    }
+    // Turn off the modem
+    network_off(index, 0, 0, NULL);
+    return resume;
+}
+
+static int system_sleep_network_resume(network_interface_index index) {
+    network_on(NETWORK_INTERFACE_ETHERNET, 0, 0, nullptr);
+    network_connect(NETWORK_INTERFACE_ETHERNET, 0, 0, nullptr);
+    return SYSTEM_ERROR_NONE;
+}
+#endif
+
+int system_sleep_ext(const hal_sleep_config_t* config, hal_wakeup_source_base_t** reason, void* reserved) {
+    SYSTEM_THREAD_CONTEXT_SYNC(system_sleep_ext(config, reason, reserved));
+
+    // Validates the sleep configuration previous to disconnecting network,
+    // so that the network status remains if the configuration is invalid.
+    CHECK(hal_sleep_validate_config(config, nullptr));
+
+    SystemSleepConfigurationHelper configHelper(config);
+    int ret;
+
+#ifndef SPARK_NO_CLOUD
+    bool cloudResume = false;
+    // Disconnect from cloud is necessary.
+    // Make sure all confirmable UDP messages are sent and acknowledged before sleeping
+    if (configHelper.cloudDisconnectRequested() && spark_cloud_flag_connected()) {
+        if (configHelper.sleepWait() == SystemSleepWait::CLOUD) {
+            Spark_Sleep();
+        }
+        cloudResume = spark_cloud_flag_auto_connect();
+        // Clear the auto connect status
+        spark_cloud_flag_disconnect();
+    }
+#endif
+
+    // Network disconnect
+#if HAL_PLATFORM_CELLULAR
```

Comment:
`if_get_list()` could potentially be used on Gen3 to traverse the list of interfaces, but let's leave for now as-is, as it allows to share to code between Gen 2 and Gen 3. Maybe add a FIXME.